### DISCUSSION TITLE
Define environment variables from kubernetes

### DISCUSSION THREAD

**Discussion by volnei** [vercel/next.js#46919] [Help] [answered]
### Summary

I just can't define variables from kubernetes after build.

next.config.js
```javascript
const API_URL = process.env.NEXT_PUBLIC_API;

...
async rewrites() {
  return [
    {
       source: '/api/:path*',
       destination: API_URL.concat('/:path*'),
   },
...

```

The variable is defined correctly in pod, but still getting it from .env.production used for build

Any tips?

**Comment by gtjamesa** [comment]
You can define non-public environment variables at runtime, but any variables prefixed with `NEXT_PUBLIC_` are resolved at build time. If you check the compiled JavaScript inside `.next/`, you'll see that the `NEXT_PUBLIC_API` string won't exist at all and was replaced by its value during build

**Reply by Sumis34** [reply]
I can not even get the normal env vars to populate at run time.

**Comment by leerob** [comment, marked as answer]
This should be solved with the App Router. We can safely read environment variables on the server during dynamic rendering. This allows you to use a singular Docker image that can be promoted through multiple environments with different values. So you could use whatever environment you want. Build once, run anywhere.

```jsx
import { unstable_noStore as noStore } from 'next/cache';
 
export default function Component() {
  noStore(); // Opt-into dynamic rendering
  // This value will be evaluated at runtime
  const value = process.env.MY_VALUE
  ...
}
```

You can also opt into dynamic rendering in other ways, like using `cookies` or `headers`, but this is an example.
